src/System.IO.Pipelines.Networking.Tls/Internal/OpenSsl/Interop.cs

```diff
@@ -0,0 +1,451 @@
+            public extern static IntPtr SSL_CIPHER_get_name(IntPtr ssl_cipher);
+            [DllImport(SslDll, CallingConvention = CallingConvention.Cdecl)]
+            public extern static IntPtr SSL_get_current_cipher(IntPtr ssl);
+
+            public static CipherInfo GetCipherInfo(IntPtr ssl)
+            {
+                var returnValue = new CipherInfo();
+                var cPtr = SSL_get_current_cipher(ssl);
+                returnValue.Name = Marshal.PtrToStringAnsi(SSL_CIPHER_get_name(cPtr));
+
+                return returnValue;
+            }
+
+            [DllImport(SslDll, CallingConvention = CallingConvention.Cdecl)]
+            public extern static IntPtr SSL_new(IntPtr sslContext);
+            [DllImport(SslDll, CallingConvention = CallingConvention.Cdecl)]
+            public extern static void SSL_CTX_free(IntPtr sslCtx);
+            [DllImport(SslDll, CallingConvention = CallingConvention.Cdecl)]
+            public extern static IntPtr SSLv23_client_method();
+            [DllImport(SslDll, CallingConvention = CallingConvention.Cdecl)]
+            public extern static IntPtr SSLv23_server_method();
+            [DllImport(SslDll, CallingConvention = CallingConvention.Cdecl)]
+            private extern static int SSL_CTX_ctrl(IntPtr ctx, int ctrlType, long options, IntPtr other);
+            public static int SSL_CTX_set_options(IntPtr ctx, ContextOptions options) => SSL_CTX_ctrl(ctx, SSL_CTRL_OPTIONS, (long)options, IntPtr.Zero);
+
+            private static readonly IntPtr ServerMethod = SSLv23_server_method();
+            private static readonly IntPtr ClientMethod = SSLv23_client_method();
+            public static IntPtr NewServerContext(VerifyMode mode)
+            {
+                var returnValue = SSL_CTX_new(ServerMethod);
+                SSL_CTX_set_options(returnValue, Default_Context_Options);
+                SSL_CTX_set_verify(returnValue, mode, IntPtr.Zero);
+                return returnValue;
+            }
+            public static IntPtr NewClientContext(VerifyMode mode)
+            {
+                var returnValue = SSL_CTX_new(ClientMethod);
+                SSL_CTX_set_options(returnValue, Default_Context_Options);
+                SSL_CTX_set_verify(returnValue, mode, IntPtr.Zero);
+                return returnValue;
+            }
+
+            public static int SetKeys(IntPtr ctx, IntPtr certificates, IntPtr privateKey)
+            {
+                if (certificates != IntPtr.Zero)
+                {
+                    InteropCrypto.CheckForErrorOrThrow(SSL_CTX_use_certificate(ctx, certificates));
+                }
+                if (privateKey != IntPtr.Zero)
+                {
+                    InteropCrypto.CheckForErrorOrThrow(SSL_CTX_use_PrivateKey(ctx, privateKey));
+                }
+                return 0;
+            }
+        }
+
+        private static class WindowsLib
+        {
+            public const string SslDll = "ssleay32.dll";
```

Comment:
> So what about this... the securechannel and openssl is the level of abstraction and that is decided automatically at runtime...

+1 Also, if you make the switch a public API, you retain the flexibility of devs switching to whatever implementation of the SPI they want/need.